docs/manifest.md

```diff
@@ -0,0 +1,104 @@
+# Manifest
+
+A Fondant manifest describes the contents of a dataset and is used as a reference to the dataset 
+to be passed between Fondant components.
+
+Creating the manifest is handled by Fondant, as each component will automatically generate the 
+manifest that represents its output dataset. While you should not have to create the manifest 
+yourself, it is still useful to understand how it works.
+
+## Contents
+
+A manifest consists of the following sections
+
+```json
+{
+  "metadata": {
+    ...
+  },
+  "index": {
+    ...
+  },
+  "subsets": {
+    ...
+  }
+}
+```
+
+### Metadata
+
+The metadata section tracks metadata about the dataset. Currently, it tracks the location where 
+the data is stored and references to the pipeline and component that were used to create it.
+
+```json
+{
+  "metadata": {
+    "base_path": "gs://bucket",
+    "run_id": "12345",
+    "component_id": "67890"
+  }
+}
+```
+
+This metadata is currently only used to build the location where the data should be stored, but 
+can in the future be extended and used for more advanced use cases such as data lineage and 
+pipeline caching.
+
+### Index
+
+The index represents which data points are part of the dataset. The index is tracked separately 
+from the data, so data points can be filtered out of a dataset without actually having to touch 
+the underlying data. The index is stored in a fixed format, which is why the manifest only 
+tracks the location where it is saved:
+
+```json
+{
+  "index": {
+    "location": "/index"
+  }
+}
+```
+
+### Subsets
+
+Each subset is a different view on the data contained in the dataset. Different subsets could 
+for instance be images, captions, and embeddings, all reflecting the same underlying data points.
+
+Each subset is stored separately to prevent expensive operations to the whole row when a single 
+subset is added or modified.
+
+Each subset contains two properties:
+- A location which points to the location where the underlying data is stored
+- A `"fields"` object which defines the columns available in the subset. Each field is defined 
+  by a name and an [Arrow data type](https://arrow.apache.org/docs/python/api/datatypes.html). 
+
+```json
+{
+  "subsets": {
+    "images": {
+      "location": "/images",
+      "fields": {
+        "data": {
+          "type": "binary"
+        },
+        "height": {
+          "type": "int32"
+        },
+        "width": {
+          "type": "int32"
+        }
+      }
+    },
+    "captions": {
+      "location": "/captions",
+      "fields": {
+        "data": {
+          "type": "binary"
+        }
+      }
+    }
+  }
+}
+```
+
+All subsets should contain at least the data points defined by the index.
```

Comment:
Question regarding this point, say that we have an image subset and we gathered different subsets linked to it (captions, aesthetic score, ...) where each subset was probably obtained in a separate component. Then we extend the images subset (e.g. laion retrieval) and the index along with it.

seed_images (image_subset)-> caption_component (add caption subset) -> aesthetic_filter (add aesthetic score subset) -> **image retrieval (add index and extend image_subset)** 

How do we handle the missing entries in the other subsets (captions, aesthetic score, ...) if we only plan on including them using subsequent components? 

... -> **image retrieval (add index and extend image_subset)** -> caption_component -> filter_component

I think this is not possible with the current framework? This question might have already been addressed and I could have missed.

Reply:
Indeed, this is not supported. There's two ways around this:
- Reorder the components in your pipeline (eg. in your example, move the image retrieval to the 2nd place)
- Work with non-linear DAGs when this is supported in the future (eg. in your example, the 2nd part of the pipeline could branch off and later merge again)